### PR TITLE
MAINT: Include `numpy._core` imports

### PR DESCRIPTION
**Pull request by mtsokol**
Hi!
`numpy.core` is becoming officially a private module and is being renamed to `numpy._core` in https://github.com/numpy/numpy/pull/24634 PR. 

It's backward and forward compatible change (**in terms of this change only** libraries compiled against 1.x will work with nightly numpy 2.0 installed and the other way around). Only a warning is emitted when accessing `numpy.core.*` with the previous name.

To avoid warnings for downstream libraries that use `pybind11`, such as scipy (such warnings are turned into errors in the CI), I propose this change: 
`pybind11/numpy.h` first tries to import with a new name `_core`. If numpy 1.x is installed it falls back to `core` name, therefore avoids emitting a warning when 2.0 is installed.

Please share your feedback!

### CODE REVIEW

**Comment by rwgk** [comment]
I got tripped up a few times by the try-this-import-then-that pattern, while working on changes to pybind11. I think making this safe requires a little more attention, with an eventual error message clearly indicating that both imports failed.

However, I have a more fundamental concern:

* If `_core` is made private, don't we undermine that effort by using the private access?

I see it's your own PR: https://github.com/numpy/numpy/pull/24634

What will be the official (non-private) way to obtain the `_ARRAY_API` and `_dtype_from_pep3118` pointers in the future?

If we continue to need access to those pointers, could you provide an official API for them?

**Comment by Skylion007** [comment]
Can we also ensure we just catch the ImportError? Do we have bindings for that exception type? If not, we may want to add it.

**Comment by rwgk** [comment]
> Can we also ensure we just catch the ImportError?

That would definitely be better.

Pointing out for completeness:

```
void **api_ptr = (void **) PyCapsule_GetPointer(c.ptr(), nullptr);
```

`if (PyErr_Occurred())` is missing after this. I.e. this will segfault a couple lines down if there is an error.

**Comment by ngoldbaum** [comment]
> What will be the official (non-private) way to obtain the _ARRAY_API and _dtype_from_pep3118 pointers in the future?

> If we continue to need access to those pointers, could you provide an official API for them?

Hi! I'm mentoring @mtsokol's Numpy 2.0 API migration project.

Can you open an issue on the NumPy issue tracker describing your needs for these two symbols? 

For `_dtype_from_pep3118`, @hoodmane was recently looking at writing a grammar for pep3118 format strings (see https://github.com/numpy/numpy/issues/24428) maybe it makes sense to have a small package on pypi that parses this format? Just spitballing, I don't have any context about why NumPy's PEP 3118 format string parser is in internals and don't know if there is some issue with making NumPy's implementation public.

For `_ARRAY_API`, this is used internally by NumPy to set up its own public C API. It looks like pybind11's usage is doing something very similar to numpy's `import_array()` function that downstream code is supposed to call to set up the C API. If there isn't a public way for a downstream project like pybind11 to get access to the NumPy C API symbols without going through the _ARRAY_API capsule, we probably need to add such a public mechanism that won't break if NumPy decides to change the way it sets up the public C API in the future.

We want to work with you to make sure you have public APIs for everything you need so you can move away from the things you're using in `_core`. One of the goals of making this change was to shake out downstream usages of numpy internals that we aren't aware of.

**Comment by rwgk** [comment]
Hi @ngoldbaum, some background and opinions:

* I wasn't involved in the implementation of pybind11/numpy.h. I just happen to be the most active pybind11 maintainer at the moment.

* IIUC pybind11/numpy.h was designed to work even if the numpy headers are not installed on the system on which the pybind11 bindings needing pybind11/numpy.h are compiled. Which is extremely convenient. But implies that it is assumed that the numpy API is set in stone and will never change. — Full disclosure: I was horrified when I first saw that around 3 years ago. Then amazed that nothing obviously broke or at least showed signs of bit rot in the meantime.

But maybe with the Numpy 2.0 API it's finally happening?

What do we do now? Taking away candy usually isn't going over well. Which brings me to a maybe radical question:

* Could we just leave pybind11/numpy.h as is?

* Add pybind11/numpy2.h and start over with the implementation, this time requiring the numpy headers? — We could work together to figure out what APIs we need.

Probably a silly question: how do you `#include` the numpy 2.0 API?

**Comment by ngoldbaum** [comment]
I think leaving behind `numpy.h` for older numpy releases and then adding a `numpy2.h` that does something nicer makes a lot of sense.

> Probably a silly question: how do you #include the numpy 2.0 API?

Not silly, the numpy docs are really bad about explaining this.

You need to do something like this in *one* compilation unit per C extension:

```
#define NPY_TARGET_VERSION NPY_2_0_API_VERSION
#include "numpy/arrayobject.h"
```

and then later in that same file do:

```
import_array();
```

(in the extension module I'm working on right now this happens inside of the `PyMODINIT_FUNC` module initialization function)

Other compilation units need something like:

```
#define NPY_TARGET_VERSION NPY_2_0_API_VERSION
#define NO_IMPORT_ARRAY
#include "numpy/arrayobject.h"
```

The `NO_IMPORT_ARRAY` bit is to ensure `import_array` only gets called once per extension.

You probably also want `#define NPY_NO_DEPRECATED_API NPY_2_0_API_VERSION` to get warnings about uses of deprecated API items, but that's not required.

**Comment by mtsokol** [comment]
> Could we just leave pybind11/numpy.h as is?

@rwgk For now I think yes, in terms of https://github.com/numpy/numpy/pull/24634 changes - I don't plan any other changes to `core` module. I will push additional checks to this PR that you mentioned in the comments (I think it can be merged to avoid warnings in CI, if you build against numpy nightly builds).
 
> Add pybind11/numpy2.h and start over with the implementation, this time requiring the numpy headers? - We could work together to figure out what APIs we need.

Sounds good in a long term - I'm curious about changes that would be required to `include/numpy.h` file.

**Comment by mtsokol** [comment]
Ok, I added additional import checks for each step.

**Comment by rwgk** [comment]
@ngoldbaum A couple thoughts:

```
#define NPY_TARGET_VERSION NPY_2_0_API_VERSION
#define NO_IMPORT_ARRAY
```

This looks like you've chosen a "maximally incremental" path. I'm not sure if my numpy2.h idea makes sense in that case, but we can see.

Is the API above (the two defines) already set in stone?

> The NO_IMPORT_ARRAY bit is to ensure import_array only gets called once per extension.

So this sounds like <nothing> is needed 1 time, `NO_IMPORT_ARRAY` is needed N times.

Did you consider something like this?

```
#define NPY_TARGET_VERSION NPY_2_0_API_VERSION
#define NPY_2_0_API_IMPORT_ARRAY_IN_THIS_TRANSLATION_UNIT
```

That way you'd have something special only 1 time, <nothing> N times.

**Comment by ngoldbaum** [comment]
> Is the API above (the two defines) already set in stone?

This is how it's always worked in NumPy, you'd just be targeting a different API version to target an older version of the NumPy API.

> Did you consider something like this?

I'm not sure if there is a problem with adding a new macro, I'd open an upstream NumPy issue.

File: include/pybind11/numpy.h
Comment (rwgk) [inline include/pybind11/numpy.h:125]: Could you please add a (terse) comment here explaining when this is expected to work, and below similarly?

Rough idea:

```
// numpy._core was established with numpy PR ...
```

```
// numpy.core was the import needed up until numpy release ...
```

File: include/pybind11/numpy.h
Comment (rwgk) [inline include/pybind11/numpy.h:140]: What you have here looks acceptable, but I wonder if we could make this more robust pretty easily, along the lines of:

* `import numpy`
* Inspect `numpy.__version__` or similar to decide if `core` or `_core` is needed, and try only that one import. (I'm hoping this might be more feasible than usual because you own the numpy API.)

This is less likely to be surprising if there is some low-level issue in pybind11 or Python itself (refactoring, new major dev version), or numpy is changed further in the future. The error message then could be the original import error, re-raised using `raise_from` (very easy):

https://github.com/pybind/pybind11/blob/7e5edbc947168fa73e25c5b6e58d656f5e2e9863/include/pybind11/pytypes.h#L817

File: include/pybind11/numpy.h
Comment (rwgk) [inline include/pybind11/numpy.h:287]: It's an existing problem, but could you please add this fix or similar while you're at it (untested)?

```
if (api_ptr == nullptr) {
    raise_from(PyExc_SystemError, "FAILURE obtaining numpy _ARRAY_API pointer.");
}
```

**Comment by rwgk** [comment]
I just added @EthanSteinberg as a reviewer. I believe he knows much more about numpy than I do.

**Comment by EthanSteinberg** [comment]
Thanks for helping fix this @ngoldbaum. I agree with @rwgk's recommendations but have two comments

1. We absolutely cannot use numpy's C code or headers because that significantly complicates the build system (which is almost always a mess for C++ extensions). So we cannot use `import_array();` or `#include "numpy/arrayobject.h"` or similar.

2. What we can do is transition away from numpy's C api and instead work with its python api.  There will probably be some slight overhead here, but I expect it will be minimal, especially because we can use tools like https://docs.python.org/3/c-api/buffer.html#c.PyObject_GetBuffer to avoid a lot of overhead.

This will require a significant revamp, but we can do it gradually, removing one numpy C api call at a time and replacing each of those calls with their python api alternatives.

What are your thoughts @rwgk, should we go through that effort? Is it worth the cost of maybe introducing bugs?

File: include/pybind11/numpy.h
Comment (mtsokol) [reply include/pybind11/numpy.h:140]: So I did similar changes for JAX and originally I also decided to go with numpy version checks in the `if` statement and then a single import.  
But I was advised that `try ... catch` approach is more robust. Here's an explanation from Jake: https://github.com/google/jax/pull/16972#pullrequestreview-1563181273

>Version checks are fine for tests, but I'd prefer to avoid them in package import paths – e.g. imagine a user who has some custom numpy build with a version string that we don't anticipate, and then import jax might fail on trying to cast the version to a tuple of ints.

But I'm happy to switch to a version check if you prefer it!

**Comment by rwgk** [comment]
> We absolutely cannot use numpy's C code or headers because that significantly complicates the build system

This makes me very sad TBH, but mainly because there is no way I could reasonably disagree.

> What we can do is transition away from numpy's C api and instead work with its python api. 

That sounds like a great plan!

> This will require a significant revamp ... should we go through that effort?

It would be fantastic if someone stepped up to do it. (I can only offer to do timely reviews at the moment.)

But it sounds like we could also incrementally work towards that goal? Like what we're currently doing here?

File: include/pybind11/numpy.h
Comment (rwgk) [reply include/pybind11/numpy.h:140]: We're dealing with 1 try vs 2 tries here, that's similar but not the same.

> version check

I'm thinking of it more as a "capability check".

(Version numbers a convenient way to answer "what exactly do I have", that's a different purpose.)

I'm still hoping that we can rise to a level of organization that allows us to only have 1 import and **report the original error**: does or can numpy provide a simple API that allows us to only try once?

**Comment by EthanSteinberg** [comment]
I can offer to help port over at least some of the calls, but not until at least November. I'll add a reminder to my calender.

File: include/pybind11/numpy.h
Comment (mtsokol) [reply include/pybind11/numpy.h:138]: Sure! I switched to an `if` statement that selects package path based on numpy version - now `(_)core.multiarray` or `(_)core._internal` is imported only once.  
(it might look lengthy, but this is recommended way to get major version integer: https://numpy.org/doc/stable/release/1.9.0-notes.html#numpyversion-class-added)

File: include/pybind11/numpy.h
Comment (mtsokol) [reply include/pybind11/numpy.h:134]: Sure! I added a comment.

File: include/pybind11/numpy.h
Comment (mtsokol) [reply include/pybind11/numpy.h:285]: Sure, I added this `if` statement - looks good to me.

**Comment by mtsokol** [comment]
I see that one of the pre-commit checks disallows `Numpy` capitalization - but it's needed to access an attribute. Can I remove that keyword from pre-commit config?

**Comment by rwgk** [comment]
```
Disallow improper capitalization.........................................Failed
- hook id: disallow-caps
- exit code: 1

include/pybind11/numpy.h:128:    object numpy_version = numpy_lib.attr("NumpyVersion")(version_string);
```

I've never seen an error like that before. It seems super weird that it would complain about caps in a string literal, how can that possibly be useful? If you have more background knowledge and know how to deal with this, please try a commit, maybe I'll understand better when I see it.

**Comment by mtsokol** [comment]
> I've never seen an error like that before. It seems super weird that it would complain about caps in a string literal, how can that possibly be useful? If you have more background knowledge and know how to deal with this, please try a commit, maybe I'll understand better when I see it.

It's `pygrep` check in pre-commit configuration that forbids "Numpy": https://github.com/pybind/pybind11/blob/7e5edbc947168fa73e25c5b6e58d656f5e2e9863/.pre-commit-config.yaml#L145

I guess it's made this way to only allow "numpy" or "NumPy" capitalization in the source code or docs. Well, here I need to access `np.lib.NumpyVersion` that was introduced at least 9 years ago so I would say "Numpy" is also legal spelling. I pushed a change that drops this constraint.

File: include/pybind11/numpy.h
Comment (rwgk) [reply include/pybind11/numpy.h:140]: (I think I somehow lost a comment here that I thought I typed in already. Sorry if something appears twice.)

```
std::string numpy_core_path = major_version >= 2 ? "numpy._core" : "numpy.core";
```

What you have looks fine though, if you prefer, the only thing I'd do is remove the explicit `std::string()` inside the two `if` branches, i.e. I think simply `numpy_core_path = "numpy._core";` will work and is what people usually do.

**Review by rwgk — APPROVED** [review]

**Comment by mtsokol** [comment]
> What you have looks fine though, if you prefer, the only thing I'd do is remove the explicit std::string() inside the two if branches, i.e. I think simply numpy_core_path = "numpy._core"; will work and is what people usually do.

Right, I changed it to a one-line variant.

I also added `throw error_already_set();` because I found it explained this way in `exceptions.rst`. Let me know if I should reverse it. Otherwise it's ready from my side!

File: include/pybind11/numpy.h
Comment (rwgk) [inline include/pybind11/numpy.h:285]: I think you're right! This needs to be here. How did you notice?

Looks like a made a mistake in #4570.

Comment (mtsokol) [reply include/pybind11/numpy.h:285]: Here in documentation: https://pybind11.readthedocs.io/en/stable/advanced/exceptions.html#chaining-exceptions-raise-from
> To do a similar thing in pybind11, you can use the py::raise_from function. It sets the current python error indicator, so to continue propagating the exception you should throw py::error_already_set().

Comment (rwgk) [reply include/pybind11/numpy.h:285]: Good catch (it completely slipped my mind), thanks a lot!

File: .pre-commit-config.yaml
Comment (rwgk) [inline .pre-commit-config.yaml:145]: From playing around for a couple minutes, I believe this will work, too:

```diff
-    entry: PyBind|Numpy|Cmake|CCache|PyTest
+    entry: PyBind|\bNumpy\b|Cmake|CCache|PyTest
```

**Comment by rwgk** [comment]
> Otherwise it's ready from my side!

I'm happy, too.

Waiting for @EthanSteinberg or @henryiii for a 2nd set of eyes.

**Review by EthanSteinberg — APPROVED** [review]
Looks good to me! Thanks again @mtsokol for the PR!

File: .pre-commit-config.yaml
Comment (henryiii) [reply .pre-commit-config.yaml:145]: I’d prefer this, personally.

File: include/pybind11/numpy.h
Comment (henryiii) [reply include/pybind11/numpy.h:135]: I can’t really check (at a conference), but what are we using the core module for? If it’s private that means NumPy is free to break us any time. Would be very nice to start moving to whatever is supported instead longer term.

Comment (henryiii) [reply include/pybind11/numpy.h:135]: Is it only our tests?

**Review by henryiii — APPROVED** [review]
Left a few comments, but not blockers.

File: include/pybind11/numpy.h
Comment (EthanSteinberg) [reply include/pybind11/numpy.h:135]: @henryiii It's private and yes NumPy can break us at any time. @rwgk and I discuss a long term solution in this thread, which is moving us to Numpy's public Python API instead of their private C API.

Comment (EthanSteinberg) [reply include/pybind11/numpy.h:135]: (Numpy has a public C API as well, but that's difficult for us to use because that would complicate our build process quite a bit).

File: .pre-commit-config.yaml
Comment (mtsokol) [reply .pre-commit-config.yaml]: Sure! Added it.

**Comment by rwgk** [comment]
For completeness:

I ran this PR through Google's global testing (internal test ID OCL:568738744:BASE:568878849:1695833342001:c72445f1).

I didn't expect issues and in fact found none.

**Comment by rwgk** [comment]
@mtsokol Could you please update the PR description and add a Suggested changelog entry?

I already restored the usual markup template, pasting most of your original description into the template, but it's no accurate anymore (doesn't match the merged implementation). If you could tweak it that would be great. Short and terse is best (IMO).

The changelog entry is usually just one sentence, just enough to give the idea for someone glancing through the list of changes for a release.

**Comment by mtsokol** [comment]
> @mtsokol Could you please update the PR description and add a Suggested changelog entry?

@rwgk Sure! I updated the PR description.

**Comment by rwgk** [comment]
I deployed this PR Google-internally yesterday. Unfortunately I had to roll it back today. We saw a number of tests timing out, certainly related to this PR. My best guess, **but totally without proof at the moment,** is deadlocking.

It'll try to get to the bottom of it asap.

File: include/pybind11/numpy.h
Comment (EthanSteinberg) [inline include/pybind11/numpy.h:647]: @rwgk I would bet this change right here is causing the deadlocking. Look at how the old code did it all in one static call so the import was only once (even if this function got called a bunch). 

The new code is importing _internal every time this function is called.

Comment (rwgk) [reply include/pybind11/numpy.h:647]: That was my first suspicion, too, but that change by itself did not help.

Then I decided to roll back and debug after production is in a healthy state again.

Sorry I should have mentioned this before. (I'm currently battling multiple fires.)

Almost certainly we should change the code here back, but I want to get to the bottom of the timeouts first. (I only have an internal reproducer at the moment. It involves running a test 100 times, ~1/3 of the tests time out.)

**Pull request merged**
